src/systems/filecoin_vm/message/_index.md

```diff
@@ -4,6 +4,47 @@ statusIcon: 🔁
 title: VM Message - Actor Method Invocation
 ---
 
+A message is the unit of communication between two actors, and thus the primitive cause of changes
+in state. A message combines:
+
+- a token amount to be transferred to be transferred from the sender to the receiver, and
+- a method with parameters to be invoked on the receiver (optional).
+
+Actor code may send additional messages to other actors while processing a received message. 
+Messages are processed synchronously: an actor waits for a sent message to complete before resuming control.
+
+The processing of a message consumes units of computation and storage denominated in gas. 
+A message's gas limit provides an upper bound on its computation. The sender of a message pays 
+for the gas units consumed by a message's execution (including all nested messages) at a
+gas price they determine.
+
+# Message validation
+
+A syntactically invalid message should not be transmitted, retained in the message pool, or
+included in a block.
+
+A syntactically valid message:
+
+- has a well-formed, non-empty `To` address ,
+- has a well-formed, non-empty `From` address, 
+- has distinct `To` and `From ` addresses,
+- has a `CallSeqNum` no less than zero,
+- has `Value` no less than zero and no greater than the total token supply (`2e9 * 1e18`), and
+- has a `MethodNum` no less than zero,
+- has non-empty `Params` only if `MethodNum` is zero,
+- has `GasPrice` no less than zero,
+- has `GasLimit` that is at least equal to the gas consumption associated with the message's serialized bytes,
+- has `GasLimit` that is no greater than the block gas limit network parameter
+- has a total serialized size no greater than 32 KiB.
+
+A syntactically valid signed message must carry a signature that verifies the payload as having
+been signed by the public key associated with the message's `From` address.
+
+There is no semantic validation (i.e. requiring access to information outside the message itself) 
+of a message that can cause a block including the message to be invalid. Every syntactically valid
+message can be included in a block and will produce a receipt from execution. However, a message
+may fail to execute to completion, in which case it will not effect the desired state change.  
```

Comment:
the above is confusing for me: given the number of failures which cause miner penalization, i considered them semantic validation issues (e.g state issues). How do you distinguish the two? And is it fair to say "every syntactically valid message can be included in a block" when such inclusion could lead to punishment?

Reply:
Your last statement: by definition!

I see what you're saying though. There might be room to simplify. The goal is to avoid any situation that requires executing the messages of an individual block outside the context of executing them as a tipset. The state to which a block's messages apply is _not known when the block is produced_. So my thinking was, if there exists some possible state to which the message would apply correctly, we can't invalidate a block for it.

Given the monotonic nature of actor creation, you might think the checks of the `To` and `From` actor validity could be moved here, and performed by a miner _against the parent state declared by the block_ (not the state the message will actually execute against) without actually executing the whole message. But that would preclude a perfectly reasonable sequence of first creating an actor and then sending a message to it in the same block. The semantic validity of the second message depends on the first, so you end up having to execute each block's messages in isolation as well as for the whole tipset, only to achieve an _approximation_ of semantic validity.

Our decision at present is to avoid the double-execution of messages. Even if we included it, we'd still have to handle non-monotonic cases like mismatched `CallSeqNum` and insufficient funds during tipset execution.

Reply:
understood. reinforces the need to define when miner is punished vs not (or that intent as explained here) imo, but makes sense to me